src/backend/xacheck.go

```diff
@@ -0,0 +1,330 @@
+
+	// If the xc.dir is already a directory, MkdirAll does nothing
+	// if the dir is one file, return err
+	if err := os.MkdirAll(xc.dir, 0744); err != nil {
+		return err
+	}
+
+	if err := xc.LoadXaCommitErrLogs(); err != nil {
+		return err
+	}
+
+	xc.wg.Add(1)
+	go func(dc *XaCheck) {
+		defer dc.wg.Done()
+		dc.xaCommitCheck()
+	}(xc)
+
+	log.Info("xacheck.init.done")
+	return nil
+}
+
+func (xc *XaCheck) addXaCommitErrLog(new *XaCommitErr) error {
+	log := xc.log
+	log.Info("xc.addXaCommitErrLog.add:+%v", new)
+
+	if _, ok := xc.retrys[new.Xaid]; ok {
+		log.Error("xacheck.addXACommitErrLog.xaid[%v].duplicate", new)
+		return errors.Errorf("xacheck.addXACommitErrLog.xaid[%v].duplicate", new.Xaid)
+	}
+
+	xc.retrys[new.Xaid] = new
+	return nil
+}
+
+// flushXaCommitErrLog is used to write the xaCommitErrlogs to the file.
+func (xc *XaCheck) flushXaCommitErrLog() error {
+	log := xc.log
+	file := path.Join(xc.dir, xacheckJSONFile)
+
+	var xaCommitErrs XaCommitErrs
+	for _, v := range xc.retrys {
+		xaCommitErrs.Logs = append(xaCommitErrs.Logs, v)
+	}
+
+	log.Info("xacheck.flush.to.file[%v].backends.conf:%+v", file, xaCommitErrs)
+	if err := config.WriteConfig(file, xaCommitErrs); err != nil {
+		log.Panicf("xacheck.flush.config.to.file[%v].error:%v", file, err)
+		return err
+	}
+	return nil
+}
+
+// WriteXaCommitErrLog is used to write the xaCommitErrLog into the xacheck file.
+func (xc *XaCheck) WriteXaCommitErrLog(txn *Txn, state string) error {
+	// if the xc is nil, it will be nothing to do
+	if xc == nil {
+		return nil
+	}
+
+	xaCommitErr := &XaCommitErr{
+		Time:     time.Now().Format("20060102150405"),
+		Xaid:     txn.xid,
+		State:    state,
+	}
+
+	xc.mu.Lock()
+	defer xc.mu.Unlock()
+	// add the xaCommitErrLog to xacheck
+	if err := xc.addXaCommitErrLog(xaCommitErr); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// TODO if the Radon crash at the moment.
+	// flush the mem to the file
+	if err := xc.flushXaCommitErrLog(); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+// commitRetryBackends
+// 1st. xa recover to check the xid is valid to some backends which need the 2nd command
+// 2nd. xa commit/rollback 'xaid' to the Backends which need commit
+func (xc *XaCheck) commitRetryBackends(query string, scatter *Scatter, xid string) (bool, error) {
+	backends := scatter.Backends()
+	log := xc.log
+
+	// if the backend is empty, output warning log.
+	if len(backends) == 0 {
+		log.Error("xacheck.commitRetryBackends.backend.empty.")
+		return false, errors.New("xacheck.backend.empty")
+	}
+
+	// the 1st stage: xa recover
+	txn, err := scatter.CreateTransaction()
+	if err != nil {
+		log.Error("xacheck.commitRetryBackends.create.transaction.error:[%v]", err)
+		return false, err
+	}
+	defer txn.Finish()
+
+	xaRecoverQuery := "xa recover"
+	var needCommitBackends []string
+	for _, backend := range backends {
+		result, err := txn.ExecuteOnThisBackend(backend, xaRecoverQuery)
+		if err != nil {
+			continue
```

Comment:
If the 'xa recover' query returns error, we can't continue, because the carsh server may not be ok, we must return error or do nothing.